### ISSUE TITLE
Mix-ins do not work for `Enum`s

### ISSUE THREAD

**Issue by kostousov-ds** [FasterXML/jackson-databind#2787]
1. create java enum like this
```
public enum SomeEnum {
    none,
    tax10,
    tax20
}
```

2. create mixin for enum
```
public enum  SomeEnumMixin {
    @JsonProperty("zero")
    none,
    @JsonProperty("TypTyp")
    tax10,
    @JsonProperty("PytPyt")
    tax20
}
```

3. register mixin via `.addMixIn(SomeEnum.class, SomeEnumMixin.class)`

4. try to deserialize sometithing

ObjectMapper throws NullPointerException
```
java.lang.NullPointerException
	at com.fasterxml.jackson.databind.introspect.AnnotatedFieldCollector._addFieldMixIns(AnnotatedFieldCollector.java:117)
	at com.fasterxml.jackson.databind.introspect.AnnotatedFieldCollector._findFields(AnnotatedFieldCollector.java:94)
	at com.fasterxml.jackson.databind.introspect.AnnotatedFieldCollector.collect(AnnotatedFieldCollector.java:48)
	at com.fasterxml.jackson.databind.introspect.AnnotatedFieldCollector.collectFields(AnnotatedFieldCollector.java:43)
	at com.fasterxml.jackson.databind.introspect.AnnotatedClass._fields(AnnotatedClass.java:366)
	at com.fasterxml.jackson.databind.introspect.AnnotatedClass.fields(AnnotatedClass.java:338)
	at com.fasterxml.jackson.databind.introspect.POJOPropertiesCollector._addFields(POJOPropertiesCollector.java:393)
	at com.fasterxml.jackson.databind.introspect.POJOPropertiesCollector.collectAll(POJOPropertiesCollector.java:322)
	at com.fasterxml.jackson.databind.introspect.POJOPropertiesCollector.getPropertyMap(POJOPropertiesCollector.java:287)
	at com.fasterxml.jackson.databind.introspect.POJOPropertiesCollector.getProperties(POJOPropertiesCollector.java:186)
	at com.fasterxml.jackson.databind.introspect.BasicBeanDescription._properties(BasicBeanDescription.java:164)
	at com.fasterxml.jackson.databind.introspect.BasicBeanDescription.findProperties(BasicBeanDescription.java:239)
	at com.fasterxml.jackson.databind.deser.BasicDeserializerFactory._findCreatorsFromProperties(BasicDeserializerFactory.java:292)
	at com.fasterxml.jackson.databind.deser.BasicDeserializerFactory._constructDefaultValueInstantiator(BasicDeserializerFactory.java:276)
	at com.fasterxml.jackson.databind.deser.BasicDeserializerFactory.createEnumDeserializer(BasicDeserializerFactory.java:1472)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createDeserializer2(DeserializerCache.java:371)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createDeserializer(DeserializerCache.java:349)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCache2(DeserializerCache.java:264)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCacheValueDeserializer(DeserializerCache.java:244)
	at com.fasterxml.jackson.databind.deser.DeserializerCache.findValueDeserializer(DeserializerCache.java:142)
	at com.fasterxml.jackson.databind.DeserializationContext.findNonContextualValueDeserializer(DeserializationContext.java:481)
	at com.fasterxml.jackson.databind.deser.BeanDeserializerBase.resolve(BeanDeserializerBase.java:497)
	at com.fasterxml.jackson.databind.deser.std.DelegatingDeserializer.resolve(DelegatingDeserializer.java:58)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCache2(DeserializerCache.java:293)
	at com.fasterxml.jackson.databind.deser.DeserializerCache._createAndCacheValueDeserializer(DeserializerCache.java:244)
	at com.fasterxml.jackson.databind.deser.DeserializerCache.findValueDeserializer(DeserializerCache.java:142)
	at com.fasterxml.jackson.databind.DeserializationContext.findRootValueDeserializer(DeserializationContext.java:491)
	at com.fasterxml.jackson.databind.ObjectMapper._findRootDeserializer(ObjectMapper.java:4669)
	at com.fasterxml.jackson.databind.ObjectMapper._readMapAndClose(ObjectMapper.java:4478)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3434)
	at com.fasterxml.jackson.databind.ObjectMapper.readValue(ObjectMapper.java:3402)
```

I've tested on 2.11.0 and 2.11.1

**Comment by cowtowncoder** [comment]
Ok I can reproduce this.

One question: aside from NPE, did this work on an earlier versions?

**Comment by cowtowncoder** [comment]
So, as per commit message, added failing test; and fixed NPE part.

However, functionality will not work as expected, most likely because of the way JDK implements `Enum`s under the hood; entries may look like fields or perhaps instance methods but are neither if I remember correctly -- so mix-in handling functionality may not be able to attach annotations as expected to them yet (class annotations work fine and are tested).

I hope to resolve the second problem too, hence leaving this issue open.

**Comment by cowtowncoder** [comment]
Ah. The problem is that unlike POJO properties that are discovered using `AnnotatedField` / `-Method` and so on -- on which mix-ins are applied -- enum names are detected directly from fields that Enum declares, by `JacksonAnnotationIntrospector`. This means mix-ins are not indeed applied.

This should be fixed but will be bit bigger undertaking as it requires changes to `AnnotationIntrospector` interface, for one.

**Comment by ianbrandt** [comment]
It looks like [2.15](https://github.com/FasterXML/jackson/wiki/Jackson-Release-2.15) will bring support for lowercasing of serialized enums per #3053, which is one use case I have for enum mix-ins. I'd rather configure the output case on a per-mix-in basis, so I'd still very much appreciate this feature.

**Comment by cowtowncoder** [comment]
It would be great to get this fixed but right now there is no good plan to do that, unfortunately.

But if anyone wants to tackle it, I'd be happy to help get PR ready.

**Comment by cowtowncoder** [comment]
@JooHyukKim Ideally it would all work through standard mix-in handling and not enum-specific handling. But then again, that may be difficult in its own way.
FWTW with POJOs `AnnotatedClass` already has all mix-ins mixed in.

I think a starting challenge is figuring out what Enum values look like, to find how mix-ins apply. 
I don't remember exactly what entries are (methods?), esp. in case of method overrides but they were not quite what I expected.

**Comment by ianbrandt** [comment]
I realized I have a need to sometimes de-snake-case and rename in addition to lowercase my enum values, so it's really exciting to see the pending PR for this. Mixins should prove a lot nicer than writing custom serializers and deserializers. 🎉

**Comment by cowtowncoder** [comment]
@jinwookh correct, that's the idea. Change existing machinery to support mix-ins for Enums in general way (both for Enum class and enum values).

Main concers/questions are just that:

1. When using overrides in enum definition, do definitions change (there's some sub-classing involved as I recall, but maybe it won't affect annotation handling)
2. From users POV, how should mix-ins look like? Do they have to create throw-away `Enum` types; or just know to specify `static` Fields with matching names? Basically, not trying to specify, say, `Method`s as those would not match (or should they? I guess with enough work it's possible but could get ugly).

**Comment by cowtowncoder** [comment]
> > ... correct, that's t
> 
> I think someone else is tagged 😅here.

Sorry. Auto-completion for the win. :-/

**Comment by JooHyukKim** [comment]
It seems this issue can be closed now, addressed by #3990 ?

**Comment by cowtowncoder** [comment]
Yes, included as fixed in 2.16.0-rc1 release notes, just forgot to close issue itself.
